### PR TITLE
Variant creation enhancement

### LINKED ISSUE #790: improve variants section on product creation
**Is your feature request related to a problem? Please describe.**
Currently product variants creation requires "save" and "queued job" before it populates.

**Describe the solution you'd like**
- doesn't require save to populate the variant table
- able to fill/edit minimal variant info (sku, price, stock, backorder, purchasablity) in base product page

example: https://youtu.be/qukulIfqtjs?t=136

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by wychoong**
closes #790 

features
- show variants immediately after selecting option values
- quick edit for variant data (sku/gtin.., price, stock, backorder)

some quirks
- option value sequence is not the same as variant's label
- on Edit, adding new variant will take first variant's price, is this a bug or 'feature' to get same price?
- on Edit, toggle an option value (eg: color->red) belongs to an existing variant will lost the variant data, and creating variant with same sku/gtin.. will hit unique validation error

image

https://user-images.githubusercontent.com/67364036/211844967-fbbf6a6e-7c10-4ecb-a929-6dab0618c161.mov

### CODE REVIEW

File: packages/admin/src/Http/Livewire/Components/Products/AbstractProduct.php [hunk 1]
```diff
@@ -526,6 +672,66 @@ public function save()
...
+            })
+            ->toArray();
+    }
+
+    public function setVariants()
+    {
+        $selectedOptionValueNames = $this->getSelectedOptionValues();
+        $selectedOptionValues = ProductOptionValue::findMany($this->optionValues)
+            ->groupBy('product_option_id');
+        ## how to sort with Options position here?
```
[the remark this answers is not in the document]
Comment (alecritson) [reply packages/admin/src/Http/Livewire/Components/Products/AbstractProduct.php:695]: @wychoong Product option values now have a `position` column so we should be able to sort using that.

[the remark this answers is not in the document]
Comment (wychoong) [reply packages/admin/src/Http/Livewire/Components/Products/AbstractProduct.php]: done

File: packages/admin/tests/Unit/Jobs/Products/GenerateVariantsTest.php [hunk 2]
```diff
@@ -21,6 +21,13 @@ class GenerateVariantsTest extends TestCase
 {
     use RefreshDatabase;
 
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $this->markTestSkipped('variant generation test should be in livewire\AbstractProduct component');
+    }
```
Comment (wychoong) [inline packages/admin/tests/Unit/Jobs/Products/GenerateVariantsTest.php:29]: this test is skipped for now and should be in AbstractProduct save

**Comment by glennjacobs** [comment]
Hi @wychoong 

Great start, I'll report back with my suggestions soon.

Thanks
Glenn

**Comment by wychoong** [comment]
hey @glennjacobs 
I noticed that i missed out customer group pricing in the new UI and would like to add that in. Do you have feedbacks on the current ones that I should improve on while making the changes to handle cg price

**Comment by glennjacobs** [comment]
> hey @glennjacobs I noticed that i missed out customer group pricing in the new UI and would like to add that in. Do you have feedbacks on the current ones that I should improve on while making the changes to handle cg price

Do you not think it would be a little too much to have Customer Group pricing at the product level?

**Comment by wychoong** [comment]
i agree, thus I'm thinking how to indicate that a variant enabled Customer Group pricing

my idea now is:
- only show the extra prices if enabled
- expand on click

**Comment by glennjacobs** [comment]
> i agree, thus I'm thinking how to indicate that a variant enabled Customer Group pricing
> 
> my idea now is:
> 
> * only show the extra prices if enabled
> * expand on click

I think if you want to update a variant's more advanced data, then you need to click through to the variant edit screen.

**Comment by wychoong** [comment]
but for a shop with large catalogue, the staff may not remember which variant have customer group pricing enabled.

for product with few variants may not be a pain to click thru each of it to change the price
but if have variants like 10+, it will be a difficult task, and when it got a mix of variants with and without CG price in 1 product will be worst to have the changes lost in the 'table' if click on variant edit page without saving.

as this ui change is to bring some ease to quick edit sku, `price`, stock... etc, i think it will be nice to have `CG price` out there as it is part of product `price`

a quick poc looks like this, i think is quite ok

https://user-images.githubusercontent.com/67364036/221172585-6f598cad-1cd6-43c5-b49f-93fdb05077ea.mov

to avoid making it too complicated, this is only for during Edit, not Create

**Comment by glennjacobs** [comment]
We still need to review this properly, but priorities have prevented it. Due to this and file conflicts, I have converted to Draft for the moment.

**Comment by glennjacobs** [comment]
@wychoong this is a nice improvement, but considering our ambition to move to Filament v3, can you let me know if you still want to be working on this update now? Happy to review and feedback if you do.

**Comment by wychoong** [comment]
> @wychoong this is a nice improvement, but considering our ambition to move to Filament v3, can you let me know if you still want to be working on this update now? Happy to review and feedback if you do.

If the plan is to use Filament's form builder, I think it will be better to wait for v3.

**Comment by glennjacobs** [comment]
> > @wychoong this is a nice improvement, but considering our ambition to move to Filament v3, can you let me know if you still want to be working on this update now? Happy to review and feedback if you do.
> 
> If the plan is to use Filament's form builder, I think it will be better to wait for v3.

OK, I will close this PR then. Thanks.

**Comment by glennjacobs** [comment]
@wychoong Due to timeframes and client commitments, we would like to continue this development. Would you like to work on it, or shall we take over the development?

**Comment by wychoong** [comment]
> @wychoong Due to timeframes and client commitments, we would like to continue this development. Would you like to work on it, or shall we take over the development?

I think would be best for you to take over the development as I’m currently working on another items.

**Comment by glennjacobs** [comment]
> > @wychoong Due to timeframes and client commitments, we would like to continue this development. Would you like to work on it, or shall we take over the development?
> 
> I think would be best for you to take over the development as I’m currently working on another items.

Thanks for confirming, will do.

**Pull request closed without merging**